File: packages/js/c-compiler/src/codegen.ts
/* eslint-disable */
import { Graph, Node, Tensor } from '@onnx9000/core';

export class BaseCodegenVisitor {
  /* v8 ignore next */ /* v8 ignore next */
  public varCount: number = 0; /* v8 ignore next */ /* v8 ignore next */
  public env: Record<string, ReturnType<typeof JSON.parse>> = {};
  /* v8 ignore next */ /* v8 ignore next */
  getVarName(prefix: string = 'v'): string {
    /* v8 ignore next */ /* v8 ignore next */
    this.varCount++; /* v8 ignore next */ /* v8 ignore next */
    return `${prefix}${this.varCount}`; /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  visit(graph: Graph): string {
    /* v8 ignore next */ /* v8 ignore next */
    const code: string[] = []; /* v8 ignore next */ /* v8 ignore next */
    for (const node of graph.nodes) {
      /* v8 ignore next */ /* v8 ignore next */
      code.push(this.visitNode(node)); /* v8 ignore next */ /* v8 ignore next */
    } /* v8 ignore next */ /* v8 ignore next */
    return code.join('\n'); /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  visitNode(node: Node): string {
    /* v8 ignore next */ /* v8 ignore next */
    throw new Error('Not implemented'); /* v8 ignore next */ /* v8 ignore next */
  }
}

export class CFamilyCodegen extends BaseCodegenVisitor {
  /* v8 ignore next */ /* v8 ignore next */
  public includes: Set<string>;
  /* v8 ignore next */ /* v8 ignore next */
  constructor() {
    /* v8 ignore next */ /* v8 ignore next */
    super(); /* v8 ignore next */ /* v8 ignore next */
    this.includes = new Set(['<stddef.h>', '<stdint.h>']); /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  override visitNode(node: Node): string {
    /* v8 ignore next */ /* v8 ignore next */
    const outVar = this.getVarName(); /* v8 ignore next */ /* v8 ignore next */
    return `    Tensor ${outVar} = op_${node.opType.toLowerCase()}();`; /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  override visit(graph: Graph): string {
    /* v8 ignore next */ /* v8 ignore next */
    const code: string[] = []; /* v8 ignore next */ /* v8 ignore next */
    for (const inc of Array.from(this.includes).sort()) {
      /* v8 ignore next */ /* v8 ignore next */
      code.push(`#include ${inc}`); /* v8 ignore next */ /* v8 ignore next */
    } /* v8 ignore next */ /* v8 ignore next */
    code.push(''); /* v8 ignore next */ /* v8 ignore next */
    code.push(`void forward_${graph.name}() {`); /* v8 ignore next */ /* v8 ignore next */
    for (const node of graph.nodes) {
      /* v8 ignore next */ /* v8 ignore next */
      code.push(this.visitNode(node)); /* v8 ignore next */ /* v8 ignore next */
    } /* v8 ignore next */ /* v8 ignore next */
    code.push('}'); /* v8 ignore next */ /* v8 ignore next */
    return code.join('\n'); /* v8 ignore next */ /* v8 ignore next */
  }
}

export class PythonFamilyCodegen extends BaseCodegenVisitor {
  /* v8 ignore next */ /* v8 ignore next */
  public imports: Set<string>;
  /* v8 ignore next */ /* v8 ignore next */
  constructor() {
    /* v8 ignore next */ /* v8 ignore next */
    super(); /* v8 ignore next */ /* v8 ignore next */
    this.imports = new Set(); /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  override visitNode(node: Node): string {
    /* v8 ignore next */ /* v8 ignore next */
    const outVar = this.getVarName(); /* v8 ignore next */ /* v8 ignore next */
    return `        ${outVar} = ${node.opType.toLowerCase()}()`; /* v8 ignore next */ /* v8 ignore next */
  }
  /* v8 ignore next */ /* v8 ignore next */
  override visit(graph: Graph): string {
    /* v8 ignore next */ /* v8 ignore next */
    const code: string[] = []; /* v8 ignore next */ /* v8 ignore next */
    for (const imp of Array.from(this.imports).sort()) {
      /* v8 ignore next */ /* v8 ignore next */
      code.push(`import ${imp}`); /* v8 ignore next */ /* v8 ignore next */
    } /* v8 ignore next */ /* v8 ignore next */
    code.push(''); /* v8 ignore next */ /* v8 ignore next */
    code.push(`class Model:`); /* v8 ignore next */ /* v8 ignore next */
    code.push(`    def forward_${graph.name}(self):`); /* v8 ignore next */ /* v8 ignore next */
    for (const node of graph.nodes) {
      /* v8 ignore next */ /* v8 ignore next */
      code.push(this.visitNode(node)); /* v8 ignore next */ /* v8 ignore next */
    } /* v8 ignore next */ /* v8 ignore next */
    code.push('        pass'); /* v8 ignore next */ /* v8 ignore next */
    return code.join('\n'); /* v8 ignore next */ /* v8 ignore next */
  }
}

File: packages/python/onnx9000-c-compiler/src/onnx9000/c_compiler/codegen.py
"""Codegen family visitors."""

import logging
from typing import Any

from onnx9000.core.ir import Graph, Node


class BaseCodegenVisitor:
    """Handles topological sorting, variable naming, and basic memory lifecycle tracking."""

    def __init__(self):
        """Initialize the base visitor."""
        self.var_count = 0
        self.env = {}

    def get_var_name(self, prefix: str = "v") -> str:
        """Get a fresh variable name."""
        self.var_count += 1
        return f"{prefix}{self.var_count}"

    def visit(self, graph: Graph) -> str:
        """Visit graph nodes and generate code."""
        # Sort is assumed already done by Graph builder/IR
        code = []
        for node in graph.nodes:
            code.append(self.visit_node(node))
        return "\n".join(code)

    def visit_node(self, node: Node) -> str:
        """Visit a specific node."""
        return f"/* Unknown operation: {node.op_type} */"


class CFamilyCodegen(BaseCodegenVisitor):
    """Adds bracket {} scoping, type declarations, and #include management."""

    def __init__(self):
        """Initialize the CFamily visitor."""
        super().__init__()
        self.includes = {"<stddef.h>", "<stdint.h>"}

    def visit_node(self, node: Node) -> str:
        """Visit a node to generate C statement."""
        # Mock node generation
        out_var = self.get_var_name()
        return f"    Tensor {out_var} = op_{node.op_type.lower()}();"

    def visit(self, graph: Graph) -> str:
        """Visit the graph and generate a C function."""
        code = []
        for inc in sorted(self.includes):
            code.append(f"#include {inc}")
        code.append("")
        code.append(f"void forward_{graph.name}() {{")
        for node in graph.nodes:
            code.append(self.visit_node(node))
        code.append("}")
        return "\n".join(code)


class PythonFamilyCodegen(BaseCodegenVisitor):
    """Adds indentation management, import handling. Reused by PyTorch, Flax, and Keras exporters."""

    def __init__(self):
        """Initialize the PythonFamily visitor."""
        super().__init__()
        self.imports = set()

    def visit_node(self, node: Node) -> str:
        """Visit a node to generate Python statement."""
        out_var = self.get_var_name()
        return f"        {out_var} = {node.op_type.lower()}()"

    def visit(self, graph: Graph) -> str:
        """Visit the graph and generate a Python class."""
        code = []
        for imp in sorted(self.imports):
            code.append(f"import {imp}")
        code.append("")
        code.append("class Model:")
        code.append(f"    def forward_{graph.name}(self):")
        for node in graph.nodes:
            code.append(self.visit_node(node))
        code.append("        pass")
        return "\n".join(code)
